src/tribler/core/utilities/slow_coro_detection.py

```diff
@@ -0,0 +1,136 @@
+import io
+import logging
+import time
+from asyncio import Handle, Task
+from threading import Lock, Thread
+
+from typing import Optional
+
+from tribler.core.utilities.main_thread_stack_tracking import get_main_thread_stack
+
+from tribler.core.utilities.main_thread_stack_tracking import main_stack_tracking_is_activated
+
+# pylint: disable=protected-access
+
+logger = logging.getLogger(__name__)
+
+SLOW_CORO_DURATION_THRESHOLD = 1.0
+WATCHING_THREAD_INTERVAL = 1.0
+
+
+class DebugInfo:
+    def __init__(self):
+        self.handle = None
+        self.start_time = None
+
+
+current = DebugInfo()
+lock = Lock()
+
+
+class SlowCoroWatchingThread(Thread):
+    """
+    A thread that detects and reports slow coroutines.
+    """
+    def run(self):
+        prev_reported_handle = None  # to detect in the loop when we have the second report of the same slow coroutine
+        while True:
+            time.sleep(WATCHING_THREAD_INTERVAL)
+            with lock:
+                handle, start_time = current.handle, current.start_time
+            if handle is not None:
+                duration = time.time() - start_time
+                if duration > SLOW_CORO_DURATION_THRESHOLD:
+                    report_freeze(current.handle, duration, first_report=(prev_reported_handle is not handle))
+                    prev_reported_handle = handle
+                else:
+                    prev_reported_handle = None
+
+
+def patched_handle_run(self: Handle):
+    """
+    Remembers the current asyncio handle object and its starting time globally, so it becomes possible
+    to access it from the separate thread and detect slow coroutines.
+    """
+    start_time = time.time()
+    with lock:
+        current.handle, current.start_time = self, start_time
+    try:
+        original_handle_run(self)
+    finally:
+        with lock:
+            assert current.handle is self, "asyncio loop should never run one handle inside another one"
+            current.handle = current.start_time = None
+
+        duration = time.time() - start_time
+        if duration > SLOW_CORO_DURATION_THRESHOLD:
+            # The coroutine step is finished successfully (without freezing), but the execution time was too long
+            report_long_duration(self, duration)
```

Comment:
This appears to be somewhat duplicative, as this report seems to replicate (I'm not completely sure) the previous logic `SlowCoroWatchingThread.run()` as they both should be triggered in the case of a slow coroutine.
`SlowCoroWatchingThread.run()` will be triggered first and `patched_handle_run()` second.

Reply:
The difference is in the following:

`patched_handle_run()` triggers after the coroutine step is finished if the step execution takes too much time. It always triggers for slow coroutine step and shows its execution time.

`SlowCoroWatchingThread.run()` checks periodically that we are not currently in the coroutine step that already took too much time but is not finished yet. This way, we can detect a freezer coroutine that never finished or a coroutine that freezes the loop until the GUI process decides to kill the Core process. In contrast to `patched_handle_run()`, it is not guaranteed that `SlowCoroWatchingThread.run()` is triggered for each coroutine, a coroutine that was just slightly longer than the limit may be missed by `SlowCoroWatchingThread.run()`.

Also, `SlowCoroWatchingThread.run()` displays the current main stack (as we are inside a running coroutine) while `patched_handle_run()` does not display the stack (as the coroutine step is already finished).